### DISCUSSION TITLE
Components in TraversalBlock

### DISCUSSION THREAD

**Discussion by rrfreimuth** [ga4gh/vrs#551] [Q&A]
Can someone point me to a use case that demonstrates the need for Allele and CPB to be components in TraversalBlock, please? It makes sense in principle but I cannot think of when I might use those types in the same construct as Adjacency.

**Comment by larrybabb** [comment]
My understanding is that @Mrinal-Thomas-Epic is working on identifying an example or two and posting them here when he can.

**Comment by Mrinal-Thomas-Epic** [comment]
Allowing Allele and CPB to be components in a traversal block allows data producers to specify which copy of a duplicated region smaller mutations occur on. For example, among the non-identical gene duplications listed in the [PharmVar CYP2D6 structural variants], you could get a different function or activity score depending on which alleles were together on a certain copy of the gene. 

![image](https://github.com/user-attachments/assets/9a4f1120-f7a5-40a4-9dd5-6700e2748928)

@d-cameron will likely be able to give better examples of how experiments or variant calling algorithms could produce outputs that we would represent with a combination of Adjacencies, Alleles and/or CPB as components in TraversalBlocks.

[PharmVar CYP2D6 structural variants]:https://a.storyblok.com/f/70677/x/169927da7a/cyp2d6_structural-variation_v3-4.pdf

**Comment by Mrinal-Thomas-Epic** [comment]
Although looking at it now, even though its useful to have that type of phasing for Alleles and CPBs, I don't know if you would ever need to associate an orientation with Alleles or CPBs. It may make more sense to move Allele and CPB up one-level to `DerivativeMolecule` and only keep `Adjacency` within the `TraversalBlock`.

The initial use case for `TraversalBlock` was just so that we could normalize the representation of adjacencies, while still being able to represent specific orientations (and combinations of orientations) in the context of a derivative molecule. For example, these two adjacencies are equivalent in isolation, but may have different meanings in the context of a derivative molecule:
```json
{
    "adj_1": {
        "type": "Adjacency",
        "adjoinedSequences": [
            {
                "type": "SequenceLocation",
                "sequenceReference": "NC_000001.10", // IRI reference
                "end": 123
            },
            {
                "type": "SequenceLocation",
                "sequenceReference": "NC_000002.11",
                "start": 456
            }
        ],
        "linker": "CCGT"
    },
    "adj_2": {
        "type": "Adjacency",
        "adjoinedSequences": [
            {
                "type": "SequenceLocation",
                "sequenceReference": "NC_000002.11",
                "start": 456
            },
            {
                "type": "SequenceLocation",
                "sequenceReference": "NC_000001.10",
                "end": 123
            }
        ],
        "linker": "ACGG"
    }
}
```

AFAIK, this concern doesn't apply to Alleles, CPB, or Terminus.

**Reply by ahwagner** [reply]
> It may make more sense to move Allele and CPB up one-level to `DerivativeMolecule` and only keep `Adjacency` within the `TraversalBlock`.

Yeah, that might be an improvement for message compactness, as the orientation information for an Allele or CPB can be inferred in the context of adjacency traversal blocks.

**Reply by Mrinal-Thomas-Epic** [reply]
I attempted to create a PR that makes this change (#554), but am looking for feedback to make sure I've updated all the correct places

**Reply by rrfreimuth** [reply]
Thanks, @Mrinal-Thomas-Epic . I'd like to think about this a bit more and may have some follow up questions for you. I'm trying to think through 1) need for explicit orientation, and 2) implications of inter-mixing phased content (Allele, CPB) with structure (Adjacency), both semantically and with consideration to situations where the extent of an Allele or CPB might be outside of flanking Adjacencies. The 2D6 use case is helpful to anchor some of that thinking.

**Reply by Mrinal-Thomas-Epic** [reply]
I just added this toy example to my MR, which may be helpful in thinking about the need for explicit orientation.

![image](https://github.com/user-attachments/assets/689eb8c7-c4d5-44b0-b311-dac5958cb501)

**Reply by larrybabb** [reply]
@ahwagner I need some assistance from you following the intention here.

If we determine that the ordered components of the `DerivativeMolecule` do not need to always be encapsulated in a `TraversalBlock` because we can infer that their orientation is `forward` (which I think is what we are saying above). Then I am confused why we wouldn't only use `TraversalBlock` for `reverse` oriented sequence spans at which point we could simplly call it `ReversalBlock`? Otherwise you can have 2 ways to specify the `forward` sequence components.

Also, I'm not sure why we would NOT move `Adjacency` to an allowed component of `DerivativeMolecule` along with the other types.

Any clarification on where we are heading with this would be helpful.

**Reply by Mrinal-Thomas-Epic** [reply]
Can you give an example of what you mean by 2 ways to specify `forward` sequence components?

I'm not opposed to the combined change of switching `TraversalBlock` to `ReversalBlock` and allowing `Adjacency` directly on `DerivativeMolecule`. The benefit is the proposed model captures the same semantics, but is slightly more concise. However, it could also be slightly more work (but not much) for implementations to handle the logic of either interacting with an `Adjacency` or `ReversalBlock` on a `DerivativeMolecule` (rather than being guaranteed to always interact with a `TraversalBlock` containing an `Adjacency`).

**Reply by larrybabb** [reply]
> Can you give an example of what you mean by 2 ways to specify forward sequence components?

@Mrinal-Thomas-Epic wouldn't it be possible to put the same elements in a `TraversalBlock` as you can in a `DerivativeMolecule` (except for `Adjacency`)?  If so then couldn't you express the same `Allele`, `CPB` and `Terminus` combination in a `DM` and wouldn't that be the same as having a `DM` with a single `TraversalBlock` that had the same combo but with a `forward` orientation? (Maybe I'm missing a key component here)

**Reply by Mrinal-Thomas-Epic** [reply]
In the old/current model:
I interpreted the [definition](https://github.com/ga4gh/vrs/blob/2.x/schema/vrs/vrs-source.yaml#L619) as only allowing a single molecular variation per `TraversalBlock`. If you had a combination of `Allele`, `CPB` and `Terminus` each one would get their own `TraversalBlock`.

In the model proposed in #554:
A combination of `Allele`, `CPB` and `Terminus` can only be specified directly on the `DerivativeMolecule` (with orientation inferred). Instances of those three classes can no longer be components of `TraversalBlock`.
